src/DssSpell.sol

```diff
@@ -57,105 +44,192 @@ contract DssSpellAction is DssAction {
+        address MCD_POT          = DssExecLib.pot();
+        address MCD_SPOT         = DssExecLib.spotter();
+        address MCD_END_OLD      = DssExecLib.end();
+        address MCD_FLIP_LINK_A  = DssExecLib.flip("LINK-A");
+        address ILK_REGISTRY_OLD = DssExecLib.reg();
+        address PIP_LINK         = DssExecLib.getChangelogAddress("PIP_LINK");
+
+        // ------------------  END  ------------------
+
+        // Set contracts in END
+        DssExecLib.setContract(MCD_END,  "vat", MCD_VAT);
+        DssExecLib.setContract(MCD_END,  "cat", MCD_CAT);
+        DssExecLib.setContract(MCD_END,  "dog", MCD_DOG);
+        DssExecLib.setContract(MCD_END,  "vow", MCD_VOW);
+        DssExecLib.setContract(MCD_END,  "pot", MCD_POT);
+        DssExecLib.setContract(MCD_END, "spot", MCD_SPOT);
+
+        // Authorize the new END in contracts
+        DssExecLib.authorize(MCD_VAT, MCD_END);
+        DssExecLib.authorize(MCD_CAT, MCD_END);
+        DssExecLib.authorize(MCD_DOG, MCD_END);
+        DssExecLib.authorize(MCD_VOW, MCD_END);
+        DssExecLib.authorize(MCD_POT, MCD_END);
+        DssExecLib.authorize(MCD_SPOT, MCD_END);
+
+        // Set wait time in END
+        Fileable(MCD_END).file("wait", EndAbstract(MCD_END_OLD).wait());
+
+        // Deauthorize the old END in contracts
+        DssExecLib.deauthorize(MCD_VAT, MCD_END_OLD);
+        DssExecLib.deauthorize(MCD_CAT, MCD_END_OLD);
+        DssExecLib.deauthorize(MCD_VOW, MCD_END_OLD);
+        DssExecLib.deauthorize(MCD_POT, MCD_END_OLD);
+        DssExecLib.deauthorize(MCD_SPOT, MCD_END_OLD);
+
+        // Deauthorize the old END from all the FLIPS
+        // Authorize the new END in all the FLIPS
+        bytes32[] memory ilks = IlkRegistryAbstract(ILK_REGISTRY_OLD).list();
+        address[] memory flips = new address[](ilks.length);
+        for (uint256 i = 0; i < ilks.length; i++) {
+            bytes32 ilk = ilks[i];
+
+            address flip = DssExecLib.flip(ilk);
+            flips[i] = flip;
+            DssExecLib.deauthorize(flip, MCD_END_OLD);
+            DssExecLib.authorize(flip, MCD_END);
+
+            try DssExecLib.removeReaderFromOSMWhitelist(IlkRegistryAbstract(ILK_REGISTRY_OLD).pip(ilk), MCD_END_OLD) {} catch {}
+            try DssExecLib.addReaderToOSMWhitelist(IlkRegistryAbstract(ILK_REGISTRY_OLD).pip(ilk), MCD_END) {} catch {}
+        }
+
+        // ------------------  ESM  ------------------
+
+        require(ESMAbstract(MCD_ESM).end() == MCD_END, "DssSpell/error-esm-end");
+
+        // Authorize new ESM to execute in new END
+        DssExecLib.authorize(MCD_END, MCD_ESM);
+
+        // Authorize new ESM to execute in VAT
+        DssExecLib.authorize(MCD_VAT, MCD_ESM);
+
+        // Make every flipper relies the MCD_ESM
+        for (uint256 i = 0; i < flips.length; i++) {
+            DssExecLib.authorize(flips[i], MCD_ESM);
+        }
+
+        // ------------------  DOG  ------------------
+
```

Comment:
add require for 'vat' + add DogAbstract dss-interface?
```suggestion
require(DogAbstract(MCD_DOG).vat() == MCD_VAT, "DssSpell/error-dog-vat");
```

Reply:
Done!